### PR TITLE
Implement state and block overrides

### LINKED ISSUE #1267: Implement State and Block Overrides for eth_call

### PR DESCRIPTION
**Pull request by rakanalh**
# Description

## Linked Issues
- Fixes #1267

### CODE REVIEW

File: crates/evm/src/rpc_helpers/mod.rs [hunk 1]
```diff
@@ -6,4 +6,60 @@ mod tracing_utils;
...
+        (None, None) => {
+            // nothing to do
+        }
+        (Some(new_account_state), None) => {
+            db.override_replace_account_storage(
+                &account,
+                new_account_state
+                    .into_iter()
+                    .map(|(slot, value)| {
+                        (U256::from_be_bytes(slot.0), U256::from_be_bytes(value.0))
```
Comment (kpp) [inline crates/evm/src/rpc_helpers/mod.rs:54]: Maybe no need to accept U256 as args in override_insert_account_storage?

File: crates/evm/src/types.rs [hunk 2]
```diff
@@ -0,0 +1,51 @@
+//! Includes the types used by the citrea-evm.
+use std::collections::BTreeMap;
+
+use reth_primitives::{Address, B256};
+use serde::{Deserialize, Serialize};
+
+/// BlockOverrides is a set of header fields to override.
+#[derive(Clone, Debug, Default, PartialEq, Eq, Serialize, Deserialize)]
+#[serde(default, rename_all = "camelCase", deny_unknown_fields)]
+pub struct BlockOverrides {
```
Comment (kpp) [inline crates/evm/src/types.rs:10]: https://reth.rs/docs/reth/rpc/types/struct.BlockOverrides.html

**Review by ercecan — APPROVED** [review]

File: crates/evm/src/types.rs (hunk 2 above)
Comment (rakanalh) [reply crates/evm/src/types.rs:10]: That exists. However, our `number` and `time` vars are u64 but theirs are U256. If i wanted to change those in our code to U256, i would have to change a LOT of places.

Comment (kpp) [reply crates/evm/src/types.rs:10]: Unfortunately our block.number must be u256 IIRC

File: crates/evm/src/types.rs [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+    ///
+    /// For `eth_callMany` this will be the block number of the first simulated block. Each
+    /// following block increments its block number by 1
+    // Note: geth uses `number`, erigon uses `blockNumber`
+    #[serde(
+        default,
+        skip_serializing_if = "Option::is_none",
+        alias = "blockNumber"
+    )]
+    pub number: Option<u64>,
```
Comment (eyusufatik) [inline crates/evm/src/types.rs:21]: one problem with using u64's here is that we'll have a hard time with how all eth jsonrpc clients are implemented. 

for things like number, all rpc args are normally hex, and u64 will have a problem with that.

alloy U64 has out-of-box ethereum rpc support take a look at that pls

**Review by eyusufatik — CHANGES_REQUESTED** [review]

File: crates/evm/src/types.rs (hunk 2 above)
Comment (rakanalh) [reply crates/evm/src/types.rs]: Ended up using the type defined in Reth as you referred and use `saturating_to` to convert to u64 as @eyusufatik suggested.

**Review by kpp — APPROVED** [review]

File: crates/evm/src/tests/call_tests.rs [hunk 4]
```diff
@@ -1478,3 +1480,131 @@ fn test_l1_fee_halt() {
...
+            &mut working_set,
+        )
+        .unwrap();
+    }
+    evm.end_soft_confirmation_hook(&soft_confirmation_info, &mut working_set);
+    evm.finalize_hook(&[99u8; 32].into(), &mut working_set.accessory_state());
+
+    // Create EVM blocks
+    for _i in 0..10 {
+        // generate 514 more blocks
```
Comment (rakanalh) [inline crates/evm/src/tests/call_tests.rs:1534]: Fix this comment later

File: crates/evm/src/query.rs [hunk 5]
```diff
@@ -495,8 +496,8 @@ impl<C: sov_modules_api::Context> Evm<C> {
         &self,
         request: reth_rpc_types::TransactionRequest,
         block_id: Option<BlockId>,
-        _state_overrides: Option<reth_rpc_types::state::StateOverride>,
-        _block_overrides: Option<Box<reth_rpc_types::BlockOverrides>>,
+        state_overrides: Option<StateOverride>,
+        block_overrides: Option<Box<BlockOverrides>>,
```
Comment (yaziciahmet) [inline crates/evm/src/query.rs:500]: why Box?

**Review by yaziciahmet — COMMENTED** [review]
LGTM. left one question

File: crates/evm/src/query.rs (hunk 5 above)
Comment (rakanalh) [reply crates/evm/src/query.rs]: Left over... good catch.

**Review by yaziciahmet — APPROVED** [review]

**Review by ercecan — APPROVED** [review]

File: Cargo.toml [hunk 6]
```diff
@@ -158,6 +158,7 @@ alloy-sol-types = { version = "0.7.7", default-features = false, features = ["js
 alloy = { version = "0.2.1", default-features = false }
 alloy-eips = { version = "0.2.1", default-features = false }
 alloy-consensus = { version = "0.2.1", default-features = false }
+alloy-serde = { version = "0.2.1", default-features = false }
```
Comment (eyusufatik) [inline Cargo.toml:161]: are you sure you are using this?

**Review by eyusufatik — APPROVED** [review]

**Pull request merged**